lib/internal/Magento/Framework/Reflection/TypeProcessor.php

```diff
@@ -521,14 +521,87 @@ public function getParamType(ParameterReflection $param)
                 $param->getDeclaringFunction()->name
             ));
         }
-        if ($type == 'array') {
+        if ($type === 'array') {
             // try to determine class, if it's array of objects
             $paramDocBlock = $this->getParamDocBlockTag($param);
             $paramTypes = $paramDocBlock->getTypes();
             $paramType = array_shift($paramTypes);
+
+            $paramType = $this->resolveFullyQualifiedClassName($param->getDeclaringClass(), $paramType);
+
             return strpos($paramType, '[]') !== false ? $paramType : "{$paramType}[]";
         }
-        return $type;
+
+        return $this->resolveFullyQualifiedClassName($param->getDeclaringClass(), $type);
+    }
+
+    /**
+     * Resolve fully qualified type name in the class alias context
+     *
+     * @param ClassReflection $sourceClass
+     * @param string $typeName
+     * @return string
+     */
+    public function resolveFullyQualifiedClassName(ClassReflection $sourceClass, string $typeName): string
+    {
+        $typeName = trim($typeName);
+
+        // Not a class, but a basic type
+        if ((strtolower($typeName[0])) === $typeName[0]) {
+            return $typeName;
+        }
+
+        // Split array suffix
+        if (preg_match('/^(.+?)\[\]$/', $typeName, $matches)) {
+            $typeName = $matches[1];
+            $arraySuffix = '[]';
+        } else {
+            $arraySuffix = '';
+        }
+
+        // If class exists, then return it
+        try {
+            new ClassReflection($typeName);
+            return $typeName . $arraySuffix;
+        } catch (\ReflectionException $e) {
+            unset($e);
+        }
+
+        // Extract alias mapping
+        $sourceFileName = $sourceClass->getDeclaringFile();
+        $aliases = [];
+        foreach ($sourceFileName->getUses() as $use) {
+            if ($use['as'] !== null) {
+                $aliases[$use['as']] = $use['use'];
+            } else {
+                $pos = strrpos($use['use'], '\\');
+
+                $aliasName = substr($use['use'], $pos + 1);
+                $aliases[$aliasName] = $use['use'];
+            }
+        }
+
+        // Resolve FQN
+        $namespace = $sourceClass->getNamespaceName();
+
+        $pos = strpos($typeName, '\\');
+        if ($pos === 0) {
+            return substr($typeName, 1);
+        }
+
+        if ($pos === false) {
+            $namespacePrefix = $typeName;
+            $partialClassName = '';
+        } else {
+            $namespacePrefix = substr($typeName, 0, $pos);
+            $partialClassName = substr($typeName, $pos);
+        }
+
+        if (isset($aliases[$namespacePrefix])) {
+            return $aliases[$namespacePrefix] . $partialClassName . $arraySuffix;
+        }
+
+        return $namespace . '\\' . $typeName . $arraySuffix;
```

Comment:
I would suggest to implement some kind of additional check for class existence before returning the final value, just to avoid possible struggles with debugging in the future.